File: python/shield/resources/sessions.py
import json
from typing import Any, Dict, List, Union

from .pii import SessionsPii


class Sessions:
    """Manage Shield sessions."""

    def __init__(self, client):
        self._client = client
        self.pii = SessionsPii(client)

    def create(self, title: str) -> Dict[str, Any]:
        """Create a new session.

        Args:
            title: Human-readable title for the session.

        Returns:
            Created session object.
        """
        return self._client._request("POST", "/sessions", json_data={"title": title})

    def retrieve(self, session_id: str) -> Dict[str, Any]:
        """Retrieve a session by ID.

        Args:
            session_id: The session ID.

        Returns:
            Session object.
        """
        return self._client._request("GET", f"/sessions/{session_id}")

    def close(self, session_id: str) -> Dict[str, Any]:
        """Close a session permanently. No new events can be appended after closing.

        Args:
            session_id: The session ID.

        Returns:
            Dict with ``id``, ``status``, and ``closed_at``.
        """
        return self._client._request("POST", f"/sessions/{session_id}/close")

    def stamp(self, session_id: str) -> Dict[str, Any]:
        """Request an RFC 3161 timestamp against the session's hash chain tip.

        Async — returns 202 immediately. Poll ``verify.session()`` and check
        ``tsa_status`` for the result.

        Args:
            session_id: The session ID.

        Returns:
            Dict with ``session_id``, ``tsa_status``, and ``message``.
        """
        return self._client._request("POST", f"/sessions/{session_id}/stamp")

    def export(self, session_id: str, format: str = "json") -> Union[List[Dict[str, Any]], bytes]:
        """Export a session in the specified format.

        Args:
            session_id: The session ID.
            format: Export format — "json" or "pdf". Defaults to "json".

        Returns:
            List of parsed NDJSON record dicts for "json" format,
            or raw bytes for "pdf".
        """
        if format == "pdf":
            return self._client._request(
                "GET",
                f"/sessions/{session_id}/export/pdf",
                raw_response=True,
            )
        # The API streams NDJSON (one JSON object per line). response.json()
        # would fail on multi-line NDJSON, so we fetch raw bytes and parse
        # each line independently.
        raw: bytes = self._client._request(
            "GET",
            f"/sessions/{session_id}/export/json",
            raw_response=True,
        )
        records = []
        for line in raw.decode("utf-8").splitlines():
            line = line.strip()
            if line:
                records.append(json.loads(line))
        return records

File: typescript/src/resources/sessions.ts
import type { ShieldSession, CreateSessionParams, ExportFormat } from "../types";
import { verifyChain, type ExportEvent, type VerifyLocalResult } from "../hashchain";
import { SessionsPii } from "./pii";

export class Sessions {
  private _request: (method: string, path: string, body?: unknown) => Promise<unknown>;
  private _rawRequest: (method: string, path: string, body?: unknown) => Promise<Response>;
  public pii: SessionsPii;

  constructor(
    request: (method: string, path: string, body?: unknown) => Promise<unknown>,
    rawRequest: (method: string, path: string, body?: unknown) => Promise<Response>,
  ) {
    this._request = request;
    this._rawRequest = rawRequest;
    this.pii = new SessionsPii(request);
  }

  async create(params: CreateSessionParams): Promise<ShieldSession> {
    return this._request("POST", "/sessions", params) as Promise<ShieldSession>;
  }

  async retrieve(id: string): Promise<ShieldSession> {
    return this._request("GET", `/sessions/${id}`) as Promise<ShieldSession>;
  }

  export(id: string, opts: { format: "json" }): Promise<Record<string, unknown>[]>;
  export(id: string, opts: { format: "pdf" }): Promise<Response>;
  async export(id: string, opts: { format: ExportFormat }): Promise<Record<string, unknown>[] | Response> {
    if (opts.format === "pdf") {
      return this._rawRequest("GET", `/sessions/${id}/export/pdf`);
    }
    // The API streams NDJSON (one JSON object per line). response.json() would
    // fail on multi-line NDJSON, so fetch raw text and parse each line independently —
    // same approach as verifyLocal().
    const resp = await this._rawRequest("GET", `/sessions/${id}/export/json`);
    const body = await resp.text();
    const records: Record<string, unknown>[] = [];
    for (const line of body.split(/\r?\n/)) {
      if (!line.trim()) continue;
      try {
        records.push(JSON.parse(line) as Record<string, unknown>);
      } catch {
        continue; // skip malformed lines
      }
    }
    return records;
  }

  async close(id: string): Promise<{ id: string; status: string; closed_at: string }> {
    return this._request("POST", `/sessions/${id}/close`) as Promise<{ id: string; status: string; closed_at: string }>;
  }

  /**
   * Request an RFC 3161 timestamp against the session's hash chain tip.
   *
   * **Async** — the server returns 202 immediately. Poll `verify.session(id)`
   * and check `tsa_status` until it becomes `"success"` or `"failed"`.
   */
  async stamp(id: string): Promise<{ session_id: string; tsa_status: string; message: string }> {
    return this._request("POST", `/sessions/${id}/stamp`) as Promise<{ session_id: string; tsa_status: string; message: string }>;
  }

  // ARCH-020: recompute every event hash client-side and re-check the
  // prev_hash linkage from genesis. Fetches the NDJSON export stream (the
  // same artifact legal discovery uses) so the bytes being verified are the
  // same bytes an auditor would replay offline. Trust-but-verify: a caller
  // who does not want to trust the server's /verify result can run this
  // locally and fail loud on any divergence.
  async verifyLocal(id: string): Promise<VerifyLocalResult> {
    const resp = await this._rawRequest("GET", `/sessions/${id}/export/json`);
    const body = await resp.text();
    const events: ExportEvent[] = [];
    for (const line of body.split(/\r?\n/)) {
      if (!line) continue;
      let parsed: { type?: string; event?: ExportEvent };
      try {
        parsed = JSON.parse(line);
      } catch {
        continue; // skip malformed lines; footer/header are always JSON
      }
      if (parsed.type === "event" && parsed.event) {
        events.push(parsed.event);
      }
    }
    return verifyChain(events);
  }
}
